src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.HillClimbing.cs

```diff
@@ -0,0 +1,385 @@
+
+                //
+                // Make sure the new thread count doesn't exceed the ThreadPool's limits
+                // 
+                newThreadCount = Math.Min(maxThreads, newThreadCount);
+                newThreadCount = Math.Max(minThreads, newThreadCount);
+
+                //
+                // Record these numbers for posterity
+                //
+                
+                // Worker Thread Adjustment stats event
+
+
+                //
+                // If all of this caused an actual change in thread count, log that as well.
+                // 
+                if (newThreadCount != currentThreadCount)
+                    ChangeThreadCount(newThreadCount, state);
+
+                //
+                // Return the new thread count and sample interval.  This is randomized to prevent correlations with other periodic
+                // changes in throughput.  Among other things, this prevents us from getting confused by Hill Climbing instances
+                // running in other processes.
+                //
+                // If we're at minThreads, and we seem to be hurting performance by going higher, we can't go any lower to fix this.  So
+                // we'll simply stay at minThreads much longer, and only occasionally try a higher value.
+                //
+                int newSampleInterval;
+                if (ratio.real < 0.0 && newThreadCount == minThreads)
+                    newSampleInterval = (int)(0.5 + _currentSampleInterval * (10.0 * Math.Max(-ratio.real, 1.0)));
+                else
+                    newSampleInterval = _currentSampleInterval;
+
+                return (newThreadCount, newSampleInterval);
+            }
+
+            private void ChangeThreadCount(int newThreadCount, State state)
+            {
+                _lastThreadCount = newThreadCount;
+                _currentSampleInterval = _randomIntervalGenerator.Next(_sampleIntervalLow, _sampleIntervalHigh);
+                double throughput = _secondsElapsedSinceLastChange > 0 ? _completionsSinceLastChange / _secondsElapsedSinceLastChange : 0;
+                LogTransition(newThreadCount, throughput, state);
+                _secondsElapsedSinceLastChange = 0;
+                _completionsSinceLastChange = 0;
+            }
+
+            private void LogTransition(int newThreadCount, double throughput, State state)
+            {
+                // TODO: Log transitions
+            }
+
+            public void ForceChange(int newThreadCount, State state)
+            {
+                if(_lastThreadCount != newThreadCount)
+                {
+                    _currentControlSetting += newThreadCount - _lastThreadCount;
+                    ChangeThreadCount(newThreadCount, state);
+                }
+            }
+
+            private static (double real, double imaginary) GetWaveComponent(double[] samples, int numSamples, double period)
+            {
```

Comment:
Would be good to add the asserts from the original sources as well. Also generally if you see any other opportunities for adding useful asserts, it would be nice to add them even though it may create some additional burden on getting things working, as it will validate your understanding of the code and potentially find issues.